eos/eos.hpp

```diff
@@ -448,6 +448,137 @@ template <typename T> class RelativisticEOS {
   Real cl_, cl2_;
 };
 
+template <typename T> class EAPRampEOS {
+  
+  // calculate alpha given a particular pressure
+  PORTABLE_FUNCTION
+  Real alpha_p(const Real P) const {
+    if (P < Pe_) {
+      return alpha0_;
+    } else if (P > Pc_) {
+      return 1.0;
+    } else {
+      return alpha0_ + (P - Pe_) / (Pc_ - Pe_) * (1.0 - alpha0_);
+    }
+  }
+  
+  template<typename F, typename ... ARGS>
+  PORTABLE_FUNCTION
+  auto alpha_impl(F func, const Real rho, ARGS... args) const {
+    constexpr const Real a_eps {1.e-12};
+    Real a {0.0};
+    Real a_old {alpha0_};
+    Real diff {0.0};
+    Real P {0.0};
+    for (int i {0}; i < 10; ++i) {
+      P = *func(a_old*rho, args...) / a_old;
+      a = this->alpha_p(P);
+      diff = fabs(a - a_old) / fabs(a + a_old) * 2.0;
+      if (diff < a_eps) break;
+      a_old = a;
+    }
+    return std::tie(a, P);
```

Comment:
Is returning a `std::pair` the way to do things these days? Not pass in a reference?

Reply:
I think `pair` supports `std::get<N>` so that is possible, but an auto function signature, a tie, and get was the lazy way and I don't think it misses any potential optimizations or copy elisions or any such thing and I didn't have to think about any concrete types.

Reply:
yeah ok. I'm still stuck in old C++-land, pre move operator, where this is a no-no. :+1: